src/OrbitSshQt/Tunnel.cpp

```diff
@@ -142,6 +143,11 @@ outcome::result<void> Tunnel::shutdown() {
     }
     case State::kSendEOF: {
       OUTCOME_TRY(channel_->SendEOF());
+      SetState(State::kWaitRemoteEOF);
+      ABSL_FALLTHROUGH_INTENDED;
+    }
+    case State::kWaitRemoteEOF: {
+      OUTCOME_TRY(channel_->WaitRemoteEOF());
```

Comment:
I believe we need a timeout here, in case the service is stuck.

Reply:
So we need to do two things: First we need to trigger another call of `shutdown` after the timeout has passed. This can be done using QTimer::singleShot right here. Next we need to store a timestamp when we start waiting and compare that against the current time to check whether the timeout already passed. So that sounds like a `std::optional<std::chrono::duration>`.

Reply:
I added a 2 second timeout for Task, ptal